compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -479,49 +516,106 @@ generateSrcFromLf thisPkgId pkgMap m = noLoc mod
+            , ideclQualified = False
+            , ideclAs = Nothing
+            , ideclHiding = Nothing
+            } :: LImportDecl GhcPs
+        | LF.unModuleName (LF.moduleName m) /= ["GHC", "CString"]
+        ]
+    -- imports needed by the module declarations
+    declImports
+     =
+        [ noLoc $
+        ImportDecl
+            { ideclExt = noExt
             , ideclSourceSrc = NoSourceText
             , ideclName =
                   noLoc $ mkModuleName $ T.unpack $ LF.moduleNameString modRef
-            , ideclPkgQual =
-                  Just $ StringLiteral NoSourceText $ unitIdFS $ getUnitId pkgRef
+            , ideclPkgQual = Nothing
             , ideclSource = False
             , ideclSafe = False
             , ideclImplicit = False
             , ideclQualified = True
             , ideclAs = Nothing
             , ideclHiding = Nothing
             } :: LImportDecl GhcPs
-        | (pkgRef@(LF.PRImport pkgId), modRef) <-
-              nubSort $ toListOf moduleModuleRef m
-        , pkgId /= thisPkgId
+        | (_unitId, modRef) <- modRefs
+        , modRef /= LF.moduleName m
+        , LF.unModuleName modRef /= ["GHC", "Prim"]
         ]
-        -- imports needed by the module declarations
+    modRefs =
+        nubSort $
+        [ (getUnitId pkg, modRef)
+        | (pkg, modRef) <- toListOf moduleModuleRef m
+        ] ++
+        (map builtinToModuleRef $
+         concat $ do
+             dataTy <- NM.toList $ LF.moduleDataTypes m
+             case LF.dataCons dataTy of
+                 LF.DataRecord fs -> map (toListOf builtinType . snd) fs
+                 LF.DataVariant vs -> map (toListOf builtinType . snd) vs
+                 LF.DataEnum _es -> pure [])
+    builtinToModuleRef = \case
+            LF.BTInt64 -> (damlPrimUnitId, translateModName intTyCon)
+            LF.BTDecimal -> (damlPrimUnitId, LF.ModuleName ["GHC", "Types"])
+            LF.BTText -> (damlPrimUnitId, LF.ModuleName ["GHC", "Types"])
+            LF.BTTimestamp -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTDate -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTParty -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTUnit -> (damlPrimUnitId, translateModName unitTyCon)
+            LF.BTBool -> (damlPrimUnitId, translateModName boolTyCon)
+            LF.BTList -> (damlPrimUnitId, translateModName listTyCon)
+            LF.BTUpdate -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTScenario -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTContractId -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTOptional -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "Prelude"])
+            LF.BTMap -> (stdlibUnitId, LF.ModuleName ["DA", "Internal", "LF"])
+            LF.BTArrow -> (damlPrimUnitId, translateModName funTyCon)
+
+    damlPrimUnitId = stringToUnitId "daml-prim"
```

Comment:
I think that is the same as `primUnitId` from `ghc-lib`?

Reply:
you're right, changed.